### ISSUE TITLE
Bug in density Laplacian calculation in nwdft module

### ISSUE THREAD

**Issue by dmejiar** [nwchemgit/nwchem#3]
Dear developers, 

I've been working with Laplacian-dependent density functionals and I found that the Laplacian of the density is not correctly computed inside xc_rhogen.F. As it is coded (lines 351-365):
 
```fortran
                      if (dolap) then
                         do mu=inizja,ifinja
                           do n=1,nq
                              ! total
                              lap(n,ii,ipert) = lap(n,ii,ipert) + 2d0*
     &                            (delchi(n,1,mu)*ffd(n,1,mu-inizja+1) +
     &                             delchi(n,2,mu)*ffd(n,2,mu-inizja+1) +
     &                             delchi(n,3,mu)*ffd(n,3,mu-inizja+1) +
     &                             heschi(n,1,mu)*ff(n,mu-inizja+1) +
     &                             heschi(n,4,mu)*ff(n,mu-inizja+1) +
     &                             heschi(n,6,mu)*ff(n,mu-inizja+1))
                           enddo
                         enddo
                       endif
                     endif
```

would be correct if the atomic loops where from ``iat=1,natoms`` and ``jat=1,natoms``, however the loops run as ``iat=1,natoms`` and ``jat=1,iat``. As a consequence, the Laplacian should be coded similar to the density gradient, at least for the terms where ``heschi`` appears. My solution was something like

```fortran
                     if(doffd) then
                       call dgemm('n','n',nq*3,nnja,nnia,1d0,
     A                      delchi(1,1,inizia),nq*3,F,nnia,0d0,ffd,nq*3)
c
c                      build tau for meta GGA
c
                       if(kske) then
                         do mu=inizja,ifinja
                           do n=1,nq
                             ttau(n,ii,ipert) = ttau(n,ii,ipert)+0.5d0*(
     &                         delchi(n,1,mu)*ffd(n,1,mu-inizja+1)+
     &                         delchi(n,2,mu)*ffd(n,2,mu-inizja+1)+
     &                         delchi(n,3,mu)*ffd(n,3,mu-inizja+1))
                           enddo
                         enddo
                       endif

                       if (dolap) then
                         do mu=inizja,ifinja
                           do n=1,nq
cc DMR/Begin
                              lap(n,ii,ipert) = lap(n,ii,ipert) + 2d0*
     &                            (delchi(n,1,mu)*ffd(n,1,mu-inizja+1) +
     &                             delchi(n,2,mu)*ffd(n,2,mu-inizja+1) +
     &                             delchi(n,3,mu)*ffd(n,3,mu-inizja+1))
cc DMR/End
                           enddo
                         enddo
                       endif
                     endif
c
                     if((what.eq.2.and.jatcur.ne.0).or.
c
c                      We need the "grad" code at zero order in the
c                      nuclear deriv case, but we can skip this part if
c                      iat is not active
c     
     O                  (what.ne.2.and.grad)) then
                       if(what.eq.0.or.what.eq.1) then
                         call xc_dchiff(nq,inizja,ifinja,
     P                        delrho(1,1,ii,ipert),delchi,
     F                        ff)
cc DMR/Begin
                         if (dolap) then
                           do mu=inizja,ifinja
                             do n=1,nq
                               lap(n,ii,ipert) = lap(n,ii,ipert) +
     &                         ff(n,mu-inizja+1)*(heschi(n,1,mu) +
     &                         heschi(n,4,mu) + heschi(n,6,mu))
                             enddo
                           enddo
                         endif
cc DMR/End
                       elseif(what.eq.2) then
                         call xc_dchiffp(nq,ipol2,inizja,ifinja,
     P                        rho,delchi,ff,
     M                        ii,jat)
                         if (grad) then
c     
c                          Compute nuclear gradient of delrho
c     
                           call xc_drhonuc(nq,ipol,inizja,ifinja,
     D                          delrho,heschi,delchi,ff,ffd,
     I                          ii,jat)
                         endif
                       endif
                     endif
c
                     if((what.eq.2.and.iatcur.ne.0).or.
     O                  (what.ne.2.and.grad)) then
c
c                      Compute delXiat(r)*Xjat(r)*Diat,jat
c                  
                       call dgemm('n','t',nq,nnia,nnja,1d0,
     A                      chi(1,inizja),nq,F,nnia,0d0,ff,nq)
                       if(what.lt.2) then
                         call xc_dchiff(nq,inizia,ifinia,
     P                        delrho(1,1,ii,ipert),delchi,
     F                        ff)
cc DMR/Begin
                         if (dolap) then
                           do mu=inizia,ifinia
                             do n=1,nq
                               lap(n,ii,ipert) = lap(n,ii,ipert) +
     &                         ff(n,mu-inizia+1)*(heschi(n,1,mu) +
     &                         heschi(n,4,mu) + heschi(n,6,mu))
                             enddo
                           enddo
                         endif
cc DMR/End
                       elseif(what.eq.2) then
                         call xc_dchiffp(nq,ipol2,inizia,ifinia,
     P                        rho,delchi,ff,
     M                        ii,iat)
                         if(grad) then
                           call dgemm('n','t',nq*3,nnia,nnja,1d0,
     A                          delchi(1,1,inizja),nq*3,F,nnia,0d0,ffd,
     +                          nq*3)
c     
c                          Compute nuclear gradient of delrho
c     
                           call xc_drhonuc(nq,ipol,inizia,ifinia,
     D                          delrho,heschi,delchi,ff,ffd,
     I                          ii,iat)

                         endif
                       endif
                     endif
```
Note that ``subroutine td_rhogen`` should be corrected as well.

Furthermore, ``grid_pscreen``  and ``grid_pcompress`` should also be called with ``heschi`` as argument so this field can be updated in the samen maner as ``chi`` and ``delchi``.

**Comment by edoapra** [comment]
Daniel,
Thank you very much for this detailed bug report.
Do you happen to have an input file that clearly show the effect of the bug?
Cheers, Edo

**Comment by dmejiar** [comment]
Edo,


Unfortunately, I do not have an example that can be run in an unmodified version of NWChem. I have made some rather large modifications to the nwdft module of version 6.6 in order to be able to compute Laplacian-dependent XC functionals selfconsistently and that's why I noted the bug (you can see the results of this work published in https://doi.org/10.1103/PhysRevA.96.052512).


I can try to run some XDM calculations with both the original and my modified version, however, I won't be able to do them until next week. Let me know if you need me to run them.


Daniel Mejia Rodriguez

Postdoctoral Associate
QTP  Department of Physics
University of Florida
2302 New Physics Building
Gainesville, FL 32611
(352) 392-9227
________________________________

**Comment by edoapra** [comment]
Daniel,
We do have a XDM example in our QA tests (dft_xdm1). In order to run it, you simply need to
cd $NWCHEM_TOP/QA
./runtests.mpi.unix dft_xdm1
1) The result of the runtests.mpi.unix script "should" tell you if the energies have changed from the reference output
1) The new output will be 
cd $NWCHEM_TOP/QA/testoutputs/dft_xdm1.out
Could you please send us the new output file and the output (that is what appears on stderr) of the runtests.mpi.unix?
Thanks a lot, Edo

**Comment by edoapra** [comment]
If you are willing to share the code you developed for  Laplacian-dependent XC functionals, we would be very interested to merge into the NWChem code.
Thanks, Edo

**Comment by edoapra** [comment]
Daniel
This is what I got in the dft_xdm1 QA test after applying your changes to xc_rhogen().
Please let me know if this is in agreement with your own observations
```
$ ./runtests.mpi.unix procs 2 dft_xdm1

 
 Running tests/dft_xdm1/dft_xdm1 
 
     cleaning scratch
     copying input and verified output files
     running nwchem (/home/edo/park/nwchem.git/bin/LINUX64/nwchem)  with 2 processors 
 
492.1u 9.9s 4:11.12 199.9% (0t+0ds+0avg+56772max)k 12200i+318232o 69pf 0swaps
     verifying output ... failed
 
Failed

$ diff testoutputs/dft_xdm1.o*nw*
7c7
< Total DFT energy = -373.31149
---
> Total DFT energy = -373.31152
22c22
< Total DFT energy = -373.31149
---
> Total DFT energy = -373.31152
24c24
< Total DFT energy = -373.31149
---
> Total DFT energy = -373.31152
```

**Comment by dmejiar** [comment]
Edo,
I found some time to run the QA test. Here are the stderr and the new output
[dft_xdm1.log](https://github.com/nwchemgit/nwchem/files/1543930/dft_xdm1.log)



      $ ./runtests.mpi.unix procs 2 dft_xdm1 | tee xdm1.log
     
      
      Running tests/dft_xdm1/dft_xdm1 
      
          cleaning scratch
          copying input and verified output files
          running nwchem (/home/dmejiar/nwchem/bin/LINUX64/nwchem)
      
     538.646u 0.802s 4:30.43 199.4% (0t+0ds+0avg+79944max)k 16i+46232o 11pf 0swaps
          verifying output ... failed
      
     Failed

     $ diff testoutputs/dft_xdm1.o*nw*
     7c7
     < Total DFT energy = -373.31149
     ---
     > Total DFT energy = -373.31154
     22,24c22,24
     < Total DFT energy = -373.31149
     < Effective nuclear repulsion energy (a.u.) 189.7760
     < Total DFT energy = -373.31149
     ---
     > Total DFT energy = -373.31154
     > Effective nuclear repulsion energy (a.u.) 189.7764
     > Total DFT energy = -373.31154
     26c26
     <           C     2.0977    -2.5416     0.0000
     ---
     >           C     2.0977    -2.5415     0.0000
     29c29
     <           N    -2.2350     2.5284     0.0000
     ---
     >           N    -2.2349     2.5284     0.0000
     36c36
     <           N     0.0001    -0.0001     0.0000
     ---
     >           N    -0.0001    -0.0001     0.0000
     38,39c38,39
     <           N     0.0000     0.0000     0.0000
     < Effective nuclear repulsion energy (a.u.) 189.7760
     ---
     >           N     0.0001     0.0000     0.0000
     > Effective nuclear repulsion energy (a.u.) 189.7764

**Comment by edoapra** [comment]
Daniel
Does your NWChem binary include changes to pscreen/compress for heschi, too?
Thanks, Edo

**Comment by dmejiar** [comment]
It does. I changed the grid_quadv0b() calls to grid_pscreen() so it has heschi and a flag to signal  laplacian calculations.

Sent from my iPhone

Postdoctoral Associate
QTP Department of Physics
University of Florida

On Dec 8, 2017, at 5:39 PM, Edoardo Apr? <notifications@github.com<mailto:notifications@github.com>> wrote:


Daniel
Does your NWChem binary include changes to pscreen/compress for heschi, too?
Thanks, Edo

-

**Comment by dmejiar** [comment]
With respect to sharing the code, I'm willing to do it. The only thing I have to check is that the NSF and DOE grants with which I'm being paid include a commitment to release code under the GPL license. Do you know how can we address this?

Sent from my iPhone

Postdoctoral Associate
QTP Department of Physics
University of Florida

On Dec 8, 2017, at 5:39 PM, Edoardo Apr? <notifications@github.com<mailto:notifications@github.com>> wrote:


Daniel
Does your NWChem binary include changes to pscreen/compress for heschi, too?
Thanks, Edo

-

**Comment by edoapra** [comment]
What version of GPL? 
I believe  that what NWChem uses (ECL 2.0) is compatible with GPLv3
https://en.wikipedia.org/wiki/Educational_Community_License#cite_note-fsf-gpl-compat-1

**Comment by edoapra** [comment]
Not quite sure if the wikipedia page provides enough details.
ECL 2.0 is a derivative of Apache 2.0. The following page discusses compatibility between Apache and GPL
https://www.apache.org/licenses/GPL-compatibility.html

**Comment by edoapra** [comment]
After reading the  Apache foundation web-page and discussing with somebody more knowledgeable on this topic, I came to the conclusion the you cannot include GPL licensed source code into ECL 2.0 license code (at least, this is my understanding for now)

**Comment by edoapra** [comment]
Please have a look at the commit I have pushed
https://github.com/nwchemgit/nwchem/commit/17d8cef44cae08125639b1518efa2859c46a7436

**Comment by dmejiar** [comment]
Edo,
Changes are ok!

With respect to the license I will check with my PI if we can use ECL instead of GPLv3. Let me talk with him and I will let you know by Monday.

Cheers,
Daniel

Sent from my iPhone

Postdoctoral Associate
QTP Department of Physics
University of Florida

On Dec 8, 2017, at 8:45 PM, Edoardo Apr? <notifications@github.com<mailto:notifications@github.com>> wrote:


Please have a look at the commit I have pushed
17d8cef<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_nwchemgit_nwchem_commit_17d8cef44cae08125639b1518efa2859c46a7436&d=DwMCaQ&c=pZJPUDQ3SB9JplYbifm4nt2lEVG5pWx2KikqINpWlZM&r=dtIkgISaf2XeYl9_S1HTgRru8Ja1vKBAuAldyMSwPmU&m=1irhF3tSa_8pkmef-sXvX5M_aGPfiHqjvLl0fxzGceo&s=AbhB9KglQStD4O9a04QkWnp87UWM0POKDG_hfWRal2I&e=>

-

**Comment by dmejiar** [comment]
Edo,

We are still trying to understand the licensure differences (and implications thereof). For example,
if we put our code in NWChem under ECL, would we then be prevented or limited in distributing equivalent code from our website under GPL?

Cheers,
Daniel

**Comment by jeffhammond** [comment]
**DISCLAIMER: I AM NOT A LAWYER.  THIS IS NOT LEGAL ADVICE.**

@dmejiar A software license is permission for those who do not own the copyright to use the software.  The copyright holder does not need a license to use their own software and is permitted to license it as many ways as they want.

There are plenty of examples of dual licensing:
- Some companies use a dual BSD-GPL license to allow for incorporation into both Linux and BSD operating systems.
- Quantum chemists have licensed their code under both GPL and commercial terms to be included in both MPQC and ORCA.
- Intel used to license TBB under both a variant of GPL and a commercial license in order to allow both use by both open-source and proprietary programs.  However, TBB now uses an Apache license.

However, dual licensing gets complicated once you have derivative works.  For example, if someone derives from your contribution to NWChem, the derivative inherits the ECL but not the GPL terms.  On the other hand, if they derive from your website, the derivative inherits the GPL terms and not the ECL terms.  The latter could not be merged into NWChem unless *all* of the copyright holders agree to change the terms from GPL to ECL.

I'd be happy to discuss licensing in detail with your mentor over email if it would help.  I've been doing open-source development for many years now in a variety domains and while working for a variety of employers.  In my experience, copyleft just prevents a lot of people from using code and provides no benefit.  The reality is that copyleft doesn't prevent anyone from using your code improperly unless you have the means to enforce it through litigation or a credible threat thereof.  Because lawyers are expensive, the only reason you'd go down that path to enforce the GPL is if the monetary damages were demonstrable or if you are the EFF and do it as a matter of principle.

**Comment by hjjvandam** [comment]
Hi Everyone,

I agree with Jeff’s statement. There is no objection to including your code under one license in NWChem and distributing it on your web-site under another license.

Best wishes,

    Huub

________________________________
Hubertus van Dam
Brookhaven National Laboratory
Mon, Fri: ISB 114 (Bldg 734)
Tue–Thu: CSI 2-165 (Bldg 725)
Tel: 631-344-6020
hvandam@bnl.gov<mailto:hvandam@bnl.gov>
www.bnl.gov/compsci/people/staff.php?q=139<https://www.bnl.gov/compsci/people/staff.php?q=139>
www.linkedin.com/in/HuubVanDam<http://www.linkedin.com/in/HuubVanDam>
orcid.org/0000-0002-0876-3294<http://www.orcid.org/0000-0002-0876-3294>

**Comment by nwchemgit** [comment]
Closing this issue.
Daniel might want to open once he is ready to contribute his code
